clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1443,12 +1442,9 @@ public void assign(Collection<TopicPartition> partitions) {
             // be no following rebalance.
             //
             // See the ApplicationEventProcessor.process() method that handles this event for more detail.
-            applicationEventHandler.add(new AssignmentChangeEvent(subscriptions.allConsumed(), time.milliseconds()));
-
-            log.info("Assigned to partition(s): {}", partitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
-
-            if (subscriptions.assignFromUser(new HashSet<>(partitions)))
-                applicationEventHandler.add(new NewTopicsMetadataUpdateRequestEvent());
+            Timer timer = time.timer(defaultApiTimeoutMs);
+            AssignmentChangeEvent assignmentChangeEvent = new AssignmentChangeEvent(timer.currentTimeMs(), calculateDeadlineMs(timer), partitions);
+            applicationEventHandler.addAndGet(assignmentChangeEvent);
```

Comment:
With this change we're now blocking on this operation completing. Is that change intentional?

Reply:
Yes, this is intentional. In `ClassicKafkaConsumer`, the change for `SubscriptionState` is synchronous. We would like to keep this behavior, so we have to block on it.